### PR TITLE
nixos/greetd: Add greetd module

### LINKED ISSUE #87603: Package request: greetd
**Project description**
greetd is a minimal and flexible login manager daemon that makes no assumptions about what you want to launch.

**Metadata**

* homepage URL: https://git.sr.ht/~kennylevinsen/greetd
* source URL: https://git.sr.ht/~kennylevinsen/greetd
* license: gplv3
* platforms: linux

### PR DESCRIPTION
**Pull request by queezle42**
###### Motivation for this change
Closes https://github.com/NixOS/nixpkgs/issues/87603

Following the [NixOS manual](https://nixos.org/manual/nixos/stable/#sec-settings-nix-representable), this module exports a `settings`-Property that generates the underlying greetd config file.

I have not added support for specific greeters, as e.g. configuring gtkgreet involves a custom sway configuration. They can later be built on top of this module.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

/marvin opt-in
/status awaiting_reviewer

### CODE REVIEW

File: nixos/modules/services/display-managers/greetd.nix [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  tty = "tty${toString cfg.vt}";
+  settingsFormat = pkgs.formats.toml {};
+in
+{
+  options.services.greetd = {
+    enable = mkEnableOption "greetd";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.greetd;
```
Comment (ymatsiuk) [inline nixos/modules/services/display-managers/greetd.nix:15]:
```suggestion
      default = pkgs.greetd.greetd;
```

File: nixos/modules/services/display-managers/greetd.nix [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      defaultText = "pkgs.greetd";
```
Comment (ymatsiuk) [inline nixos/modules/services/display-managers/greetd.nix:16]:
```suggestion
      defaultText = "pkgs.greetd.greetd";
```

File: nixos/modules/services/display-managers/greetd.nix [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+      description = "The greetd package that should be used.";
+    };
+
+    settings = mkOption {
+      type = settingsFormat.type;
+      example = literalExample ''
+        {
+          default_session = {
+            command = "''${pkgs.greetd}/bin/agreety --cmd sway";
```
Comment (ymatsiuk) [inline nixos/modules/services/display-managers/greetd.nix:25]:
```suggestion
            command = "''${pkgs.greetd.greetd}/bin/agreety --cmd sway";
```

File: nixos/modules/services/display-managers/greetd.nix [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+          "plymouth-quit-wait.service"
+          "getty@${tty}.service"
+        ];
+        Conflicts = [
+          "getty@${tty}.service"
+        ];
+      };
+
+      serviceConfig = {
+        ExecStart = "${pkgs.greetd}/bin/greetd --config ${settingsFormat.generate "greetd.toml" cfg.settings}";
```
Comment (ymatsiuk) [inline nixos/modules/services/display-managers/greetd.nix:83]:
```suggestion
        ExecStart = "${pkgs.greetd.greetd}/bin/greetd --config ${settingsFormat.generate "greetd.toml" cfg.settings}";
```

**Review by ymatsiuk — CHANGES_REQUESTED** [review]

File: nixos/modules/services/display-managers/greetd.nix [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      description = "The greetd package that should be used.";
+    };
+
+    settings = mkOption {
+      type = settingsFormat.type;
+      example = literalExample ''
+        {
+          default_session = {
+            command = "''${pkgs.greetd}/bin/agreety --cmd sway";
+            # you have to create the 'greeter'-user
```
Comment (ymatsiuk) [inline nixos/modules/services/display-managers/greetd.nix:26]:
```suggestion
```

File: nixos/modules/services/display-managers/greetd.nix [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        Restart = "always";
+      };
+
+      # Don't kill a user session when using nixos-rebuild
+      restartIfChanged = false;
+
+      wantedBy = [ "graphical.target" ];
+    };
+
+    systemd.defaultUnit = "graphical.target";
```
Comment (ymatsiuk) [inline nixos/modules/services/display-managers/greetd.nix:100]:
```suggestion
    systemd.defaultUnit = "graphical.target";
    
    users.users."${cfg.settings.default_session.user}".isSystemUser = true;
```

**Review by ymatsiuk — CHANGES_REQUESTED** [review]

**Comment by ymatsiuk** [comment]
@qzle thanks for taking care of this!
With all the suggested changes my config looks like this:
```nix
  services.greetd = {
    enable = true;
    restart = false;
    settings = {
      default_session = {
        command = "${pkgs.greetd.tuigreet}/bin/tuigreet --time --cmd ${swayRun}";
        user = "greeter";
      };
      initial_session = {
        command = "${swayRun}";
        user = "ymatsiuk";
      };
    };
  };
```
and works perfectly as I would expect it to.

File: nixos/modules/services/display-managers/greetd.nix (hunk 1 above)
Comment (qzle) [reply nixos/modules/services/display-managers/greetd.nix]: Thanks for catching that! My own greetd-package was still active while testing the module.

File: nixos/modules/services/display-managers/greetd.nix (hunk 6 above)
Comment (qzle) [reply nixos/modules/services/display-managers/greetd.nix:100]: I don't think we should mess with an unknown user. The default_session.user might be a normal user when using greetd for autologin (in this case "repeated" autologin, e.g. in a kiosk-style deployment)

**Comment by qzle** [comment]
While that works great for the common use case, the suggested change breaks with "kiosk"-style usage by changing a normal user to a system user. The simplest example that breaks (I actually use this to experiment with my pinephone):
```nix
services.greetd = {
  enable = true;
  settings = {
    default_session = {
      command = "sway";
      user = "queezle";
    };
  };
};
```

An alternative would be to create the `greeter`-user by default and set
`services.greetd.settings.default_session.usert = mkDefault "greeter";`

You could then remove the `user = "greeter";` line from your configuration; the disadvantage is that this reserves the `greeter`-username NixOS-wide.

**Comment by qzle** [comment]
I have implemented the `greeter`-user. This reduces the minimal configuration, while still preserving all configuration options, so I think this is a useful addition.

Is the username `greeter` acceptable (it comes from from the greetd documentation) or should it be namespaced to greetd, e.g. `greetd-greeter`?

File: nixos/modules/services/display-managers/greetd.nix [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+      serviceConfig = {
+        ExecStart = "${pkgs.greetd.greetd}/bin/greetd --config ${settingsFormat.generate "greetd.toml" cfg.settings}";
+
+        # Defaults from greetd upstream configuration
+        IgnoreSIGPIPE = false;
+        SendSIGHUP = true;
+        TimeoutStopSec = "30s";
+        KeyringMode = "shared";
+      } // optionalAttrs cfg.restart {
```
Comment (qzle) [inline nixos/modules/services/display-managers/greetd.nix:89]: I just remembered that https://github.com/NixOS/nixpkgs/pull/112819 was specifically for _this module_

**Review by ymatsiuk — APPROVED** [review]

**Comment by ymatsiuk** [comment]
@qzle I like your suggestion even more. Thanks for the effort 🙂

File: nixos/modules/services/display-managers/greetd.nix [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+    vt = mkOption  {
+      type = types.int;
+      default = 1;
+      description = ''
+        The virtual console (tty) that greetd should use. This option also disables getty on that tty.
+      '';
+    };
+
+    restart = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/display-managers/greetd.nix:43]: Does it make more sense to automatically do this for the sysadmin?

`serviceConfig.Restart = mkDefault cfg.settings.initial_session or false;`

**Review by zimbatm — APPROVED** [review]
LGTM, tested on my machine with sway.

File: nixos/modules/services/display-managers/greetd.nix (hunk 8 above)
Comment (qzle) [reply nixos/modules/services/display-managers/greetd.nix:43]: I like the general idea but dislike the hidden behavior, so I left the `restart`-option in but set it's default accordingly. Is that an acceptable solution?

Comment (aanderse) [reply nixos/modules/services/display-managers/greetd.nix:43]: I was really asking a question, not implying the answer... hence any solution, including the original, is an acceptable solution to me.

😄

Comment (qzle) [reply nixos/modules/services/display-managers/greetd.nix:43]: I guess that wasn't phrased well, I just wanted to ask for feedback on leaving the option :)
Anyway, I'm happy with the current version. Adding the default reduces the chance for accidental insecure deployments.

**Comment by zimbatm** [comment]
@qzle are you ready to merge or are there more things you want to address?

**Comment by qzle** [comment]
I'm done, just waiting for more reviews or a merge.

**Review by ymatsiuk — APPROVED** [review]

**Comment by bqv** [comment]
Ooh. Thanks folks

**Pull request merged**